### ISSUE TITLE
Halting development of pthreads for 7.4

### ISSUE THREAD

**Issue by krakjoe** [krakjoe/pthreads#929]
Hi all,

I'm aware that this is going to annoy some people, so apologies before we begin ...

I do not intend to maintain pthreads anymore, nor encourage it's maintenance: It is, in my opinion, broken by design, and fixing it, is simply not worth it.

The internals world is changing fast and pthreads is not going to be able to fit into that world. It is highly complicated, too difficult to deploy and understand, and slow, necessarily so because of the API it tries to provide. When we get a JIT, we will be in a position where it is much faster to execute your CPU bound code in one normal thread, than distributing that work across many pthreads. There is nothing that can be done to help pthreads while it has the API it has, it's possible that the way pthreads works will actually *prohibit the JIT from compiling code*. Everything about pthreads is putting it in an untenable position for the future.

I've started work on a new extension that will be compatible with the future: https://github.com/krakjoe/parallel

While it is not compatible with pthreads, and will not be, I view it as the only viable replacement for the versions of PHP to come, and my focus has to be on that.

When 7.4 goes into RC stage near the end of this year, this repository will be archived, and no further releases of pthreads will occur on PECL or from this repository.

Thanks for all the effort that contributors have put into pthreads, you've worked hard and I appreciate that, and I would ask that anyone with a focus on pthreads internals today shifts that focus to the future and parallel.

**Comment by dktapps** [comment]
I've been looking at parallel recently and while it looks nice, it can't replace pthreads in its current state due to a few things missing - most notably on my part I need thread-safe queues which parallel doesn't appear to have any way to create.

It appears that parallel would fit in well for a pool and task style architecture in its current form. However, for long-life threads like PocketMine's RakLib library needs, we need thread safe ways to transfer data without closure returns, since there is a persistent context which has to stick around for the duration of the runtime, and also because it's active for the whole time period.

I assume there is more planned for parallel?

**Comment by rthrfrd** [comment]
Thanks and good luck! Limiting closures sounds like a good approach. We applied the same restrictions on them at a PHP level to be able to use them in conjunction with threads - 'compiling' closures into __invokable classes prevented the use of features that made closures un-threadable.

**Comment by krakjoe** [comment]
@dktapps very much unfinished, you can see from commits it has existed for ~four days, although I've been considering it for a long time, I really only just got down to code.

The next step is something like Go's channels for bi-directional communication, I've already begun researching this (by reading the Go source), and discussing with other internals developers. Unfortunately the Go source itself is written in Go, a language I'm not overly familiar with, but for every hour I get to spend reading, the picture becomes clearer in my mind. I'm certain at this point it is doable. I've also discovered while reading, that goroutines, in architecture, are not that different to what we have in parallel - they are *mostly* share nothing, the only time a go routine reads/writes the stack of another routine is during unbuffered channel read/write operations. They are otherwise what we would call share nothing. We can emulate this very well proven, and very well liked model very well, and at sometime in the future the argument that go (or nearly anything) is better than php for parallel concurrency will all but disappear. These are not things that will ever be possible with the pthreads code base, and I'm quite sure you see the value in this.

You can dispatch a Closure to a parallel runtime, and if it doesn't return, it can persist for the duration of the thread, given channels, it can also communicate with other threads and tasks.

**Comment by beberlei** [comment]
@dktapps have you looked into using a message queue (Redis, Gearman, Beanstalkd, RabbitMQ, ...) and separating controller and worker on a per process basis?

**Comment by dktapps** [comment]
@beberlei I haven't, but that's because pthreads was here making it easy not to need multiple processes. :)

**Comment by CharlotteDunois** [comment]
Unfortunately parallel is not a replacement for pthreads. Parallel seems to be a replacement for simple tasks, but in its state it's no replacement for what you can achieve with pthreads today.

**Comment by krakjoe** [comment]
It's not intended to be a 1:1 replacement, that wouldn't make any sense, that would burden it with the same problems that pthreads has.

It is a new approach to the problem, one that is forward compatible with what is coming to PHP, one that is stable very quickly, one that many more people will be able to deploy than can deploy pthreads.

**Comment by tpunt** [comment]
Hey Joe!

I agree that `pthreads` isn't really salvageable due to the abilities it attempts to provide, causing horrible complexities (bother internally and externally), along with negative performance impacts - hence my recent lack of activity here, along with my efforts with `pht`. Whilst I believe `pht` still remains a good solution to multithreading in PHP (since it avoids mostly all internal complexities `pthreads` faces, but still provides the same power and flexibility), I also believe that `parallel` is an interesting exploration worth pursuing. I will, no doubt, be looking to contribute to it in future :)

Cheers,
Tom

**Comment by joeyhub** [comment]
I've done a fair amount of work with parallel libraries. Performance is as important as usability for a lot of people.

I'd say pthreads is definitely a mess and more complex than it strictly needs to be with various design issues.

Though over simplifying things can be complex as well.

Using an entry function definitely takes out the horrors of doing strange stuff with a class and starting with a class. The natural main scope for a new php instance is not as a method.

The future isn't good. Instead take a callback. run(Closure threaded, ?Closure onFinish).

People can put their Promise.resolve or whatever for on finish or wrap it with an event handler. I think it's best on user space to do what people want for promises and event management.

Basically get rid of he future and just take a callback instead then it's simpler, better and doesn't do more than it needs to do.

**Comment by krakjoe** [comment]
Just to be clear, these ideas about callbacks don't make any sense at all.

I'm closing this because its not meant to be a discussion, it was meant to be an announcement, and I'm quite sure anyone reading doesn't want to read a lot of nonsense.
